File: Chris/2022-07-15-NumberOfIslands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        
        
        rows, cols = len(grid), len(grid[0])
        visit = set()
        islands = 0
        
        
        def bfs(r,c):
            q = collections.deque()
            visit.add((r,c))
            q.append((r,c))
            
            while q:
                row, col = q.popleft()
                directions = [[1,0],[-1,0],[0,1],[0,-1]]
                for dr, dc in directions:
                    r, c = row + dr, col + dc
                    
                    if (r in range(rows) and c in range(cols) and grid[r][c] == "1" and (r,c) not in visit):
                        q.append((r,c))
                        visit.add((r,c))
                        
                
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == "1" and (r,c) not in visit:
                    bfs(r,c)
                    islands += 1
                            
        return islands
        
        
        

File: Chloe Gwon/2022-07-15-Number-of-Islands.java
class Solution {
    public int numIslands(final char[][] grid) {       
    int count = 0;
    for (int i=0;i<grid.length;i++)
        for (int j=0;j<grid[0].length;j++)
            if (paint(grid,i,j)) count++;
    return count;
    }
    
    final public boolean paint(final char[][] grid, final int i, final int j){
        if (i<0||i>=grid.length||j<0||j>=grid[0].length||grid[i][j]=='0') {
            return false; 
        }
        
        grid[i][j]='0';
        paint(grid,i,j+1);
        paint(grid,i,j-1);
        paint(grid,i+1,j);
        paint(grid,i-1,j);
        return true;
    }
}
